src/Api.php

```diff
@@ -0,0 +1,277 @@
+            $this->requestData = json_decode($this->request->getBody()->getContents(), true);
+        } else {
+            $this->requestData = $this->request->getParsedBody();
+        }
+
+        // This is how we will send responses
+        $this->emitter = new \Zend\Diactoros\Response\SapiEmitter();
+    }
+
+    /**
+     * Do pattern matching.
+     *
+     * @param string   $pattern
+     * @param callable $callable
+     *
+     * @return mixed
+     */
+    public function match($pattern, $callable = null)
+    {
+        $path = explode('/', rtrim($this->path, '/'));
+        $pattern = explode('/', rtrim($pattern, '/'));
+
+        $vars = [];
+
+        // $sanity = 50; unused variable
+
+        while ($path || $pattern) {
+            $p = array_shift($path);
+            $r = array_shift($pattern);
+
+            if ($p === null && $r === '') {
+                continue;
+            }
+
+            // must make sure both match
+            if ($p === $r) {
+                continue;
+            }
+
+            // pattern 'r' accepts anything
+            if ($r == '*' && strlen($p)) {
+                continue;
+            }
+
+            if ($r === null || $r === '') {
+                return false;
+            }
+
+            if ($r[0] == ':' && strlen($p)) {
+                $vars[] = $p;
+                continue;
+            }
+
+            // good until the end
+            if ($r == '**') {
+                break;
+            }
+
+            return false;
+
+            /* never used code
+            if (!$sanity--) {
+                throw new Exception(['Insanity while matching']);
+            }
+            */
+        }
+
+        // if no callable function set - just say that it matches
+        if ($callable === null) {
+            return true;
+        }
+
+        // try to call callable function
+        try {
+            $ret = call_user_func_array($callable, $vars);
+        } catch (\Exception $e) {
+            $this->caughtException($e);
+        }
+
+        // if callable function returns agile data model, then export it
+        // this is important for REST API implementation
+        if ($ret instanceof \atk4\data\Model) {
+            $ret = $ret->export();
+        }
+
+        // create response object
+        if ($ret !== null) {
+            if (!$this->response) {
+                $this->response =
+                    new \Zend\Diactoros\Response\JsonResponse(
+                        $ret,
+                        200,
+                        [],
+                        JSON_PRETTY_PRINT | JSON_HEX_TAG | JSON_HEX_APOS | JSON_HEX_AMP | JSON_HEX_QUOT
+                    );
+            }
+
+            if ($this->emitter) {
+                // cannot do anything about it
+                $this->emitter->emit($this->response);
+            }
+
+            // no emitter (is that possible at all ?)
+            return $ret;
```

Comment:
is it possible at all to not have emitter defined? it's always defined in constructor.
in what case i exclusively would want to set `$api->emitter = false` like in test script?

Reply:
yes, if we don't want response to be sent out through echo. That's a valid situation when we doing unit tests.